README.md
# Colissimo Fuel Adjustment Coefficients fetcher

Retrieve current air and road rates applied to Colissimo shipping costs and
convert them to multipliers we can directly use in your app to calculate
precise shipping costs?

## Usage

Install the gem:

    $ gem install colissimo_fuel_adjustment_coefficients

Run the gem:

    $ colissimo_fuel_adjustment_coefficients

Get the result:

    # Fuel multipliers for 05-2017
    # Fetched from https://www.colissimo.entreprise.laposte.fr/fr/system/files/imagescontent/docs/indice_gazole.xml

    AIR_FUEL_MULTIPLIER  = 1.0000
    ROAD_FUEL_MULTIPLIER = 1.0077

Or use it in your app:

    require 'colissimo_fuel_adjustment_coefficients'

    colissimo = ColissimoFuelAdjustmentCoefficients.new
    colissimo.road_multiplier      # 0.10077e1
    colissimo.air_multiplier       # 0.1e1


CHANGELOG.md
# Change Log

## [v0.3.0](https://github.com/levups/colissimo_fuel_adjustment_coefficients/tree/v0.3.0) (2018-11-19)
[Full Changelog](https://github.com/levups/colissimo_fuel_adjustment_coefficients/compare/v0.2.2...v0.3.0)

**Closed issues:**

- Replace Oga usage with plain standard REXML parsing? [\#13](https://github.com/levups/colissimo_fuel_adjustment_coefficients/issues/13)

**Merged pull requests:**

- Replace oga with REXML [\#14](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/14) ([bobmaerten](https://github.com/bobmaerten))

## [v0.2.2](https://github.com/levups/colissimo_fuel_adjustment_coefficients/tree/v0.2.2) (2018-11-16)
[Full Changelog](https://github.com/levups/colissimo_fuel_adjustment_coefficients/compare/v0.2.1...v0.2.2)

## [v0.2.1](https://github.com/levups/colissimo_fuel_adjustment_coefficients/tree/v0.2.1) (2018-11-16)
[Full Changelog](https://github.com/levups/colissimo_fuel_adjustment_coefficients/compare/v0.2.0...v0.2.1)

**Closed issues:**

- Sign gem before publishing it [\#11](https://github.com/levups/colissimo_fuel_adjustment_coefficients/issues/11)
- Comply to `bundle gem` defaults [\#10](https://github.com/levups/colissimo_fuel_adjustment_coefficients/issues/10)

**Merged pull requests:**

- Prepare for official gem release [\#12](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/12) ([bobmaerten](https://github.com/bobmaerten))

## [v0.2.0](https://github.com/levups/colissimo_fuel_adjustment_coefficients/tree/v0.2.0) (2018-11-14)
[Full Changelog](https://github.com/levups/colissimo_fuel_adjustment_coefficients/compare/v0.1.0...v0.2.0)

**Closed issues:**

- Make coefficients results usable as a BigDecimal value [\#8](https://github.com/levups/colissimo_fuel_adjustment_coefficients/issues/8)
- Add tests to the gem [\#7](https://github.com/levups/colissimo_fuel_adjustment_coefficients/issues/7)
- Move from httparty to HTTPrb for HTTP requests [\#6](https://github.com/levups/colissimo_fuel_adjustment_coefficients/issues/6)

**Merged pull requests:**

- Replace httparty with HTTP.rb, unify code and add tests [\#9](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/9) ([bobmaerten](https://github.com/bobmaerten))
- Bump httparty from 0.16.2 to 0.16.3 [\#5](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/5) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump ast from 2.3.0 to 2.4.0 [\#4](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/4) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump addressable from 2.5.1 to 2.5.2 [\#3](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/3) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump oga from 2.10 to 2.15 [\#2](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/2) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump httparty from 0.15.5 to 0.16.2 [\#1](https://github.com/levups/colissimo_fuel_adjustment_coefficients/pull/1) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v0.1.0](https://github.com/levups/colissimo_fuel_adjustment_coefficients/tree/v0.1.0) (2017-05-23)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

test/colissimo_fuel_adjustment_coefficients_test.rb
# frozen_string_literal: true

require 'date'
require 'minitest/autorun'
require 'minitest/stub_any_instance'
require 'colissimo_fuel_adjustment_coefficients'

class ColissimoFuelAdjustmentCoefficientsTest < Minitest::Test
  def test_time_period
    nominal_case do
      assert_equal '11-2018', @colissimo.time_period
    end

    failing_case do
      assert_nil @colissimo.time_period
    end
  end

  def test_air_multiplier
    nominal_case do
      assert_equal 1.0301, @colissimo.air_multiplier
    end

    failing_case do
      assert_nil @colissimo.air_multiplier
    end
  end

  def test_road_multiplier
    nominal_case do
      assert_equal 1.0197, @colissimo.road_multiplier
    end

    failing_case do
      assert_nil @colissimo.road_multiplier
    end
  end

  def test_live_values
    live_colissimo = ColissimoFuelAdjustmentCoefficients.new
    colissimo_date = Date.parse live_colissimo.time_period

    assert_equal Date.today.month, colissimo_date.month
    assert_equal Date.today.year, colissimo_date.year

    assert_kind_of BigDecimal, live_colissimo.air_multiplier
    assert_operator live_colissimo.air_multiplier, :>=, 1.0

    assert_kind_of BigDecimal, live_colissimo.road_multiplier
    assert_operator live_colissimo.road_multiplier, :>=, 1.0
  end

  private

  def sample_response
    <<~XML
      <indice_gazole>
        <!-- Tableau indice CAP -->
        <cap>
          <cap_mois>
            <titre>11-2018</titre>
            <routier>1,97%</routier>
            <aerien>3,01%</aerien>
          </cap_mois>
          <cap_mois>
            <titre>10-2018</titre>
            <routier>1,79%</routier>
            <aerien>2,33%</aerien>
          </cap_mois>
        </cap>
      </indice_gazole>
    XML
  end

  def nominal_case
    ColissimoFuelAdjustmentCoefficients.stub_any_instance(:response,
                                                          sample_response) do
      @colissimo = ColissimoFuelAdjustmentCoefficients.new
      yield
    end
  end

  def failing_case
    ColissimoFuelAdjustmentCoefficients.stub_any_instance :response, '' do
      @colissimo = ColissimoFuelAdjustmentCoefficients.new
      yield
    end
  end
end


lib/colissimo_fuel_adjustment.rb
# frozen_string_literal: true
require "addressable/uri"
require "httparty"
require "oga"
require "plain_http_client"

# <indice_gazole>
# 	<cap>
# 		<cap_mois>
# 			<titre>05-2017</titre>
# 			<routier>0,77%</routier>
# 			<aerien>0,00%</aerien>
# 		</cap_mois>
#     ...
#   </cap>
# <indice_gazole>
class ColissimoFuelAdjustment
  URL = 'https://www.colissimo.entreprise.laposte.fr/fr/system/files/imagescontent/docs/indice_gazole.xml'

  def initialize
    response = PlainHttpClient.get(url, limit: 3, timeout: 10)
    document = Oga.parse_xml response.body
    @cap = document.xpath("indice_gazole/cap/cap_mois").first
  end

  def time_period
    @cap.at_xpath("titre").text
  end

  def air_multiplier
    format_multiplier @cap.at_css("aerien").text.tr("%", "").tr(",", ".").to_f
  end

  def road_multiplier
    format_multiplier @cap.at_css("routier").text.tr("%", "").tr(",", ".").to_f
  end

  def url
    URL
  end

private

  def format_multiplier(coefficient)
    "%.4f" % (coefficient / 100 + 1)
  end
end


lib/colissimo_fuel_adjustment_coefficients.rb
# frozen_string_literal: true

require "bigdecimal"
require "http"
require "rexml/document"

# Excepted XML is like:
#   <indice_gazole>
#     <cap>
#       <cap_mois>
#         <titre>05-2017</titre>
#         <routier>0,77%</routier>
#         <aerien>0,00%</aerien>
#       </cap_mois>
#       ...
#     </cap>
#   <indice_gazole>
class ColissimoFuelAdjustmentCoefficients
  include REXML

  def initialize
    @parsed_values = {}
    return unless xml_root

    caps  = xml_root.elements.first
    month = caps.elements.first
    month.elements.each do |element|
      @parsed_values[element.name] = element.text
    end
  end

  def url
    "https://www.colissimo.entreprise.laposte.fr/fr/system/files/imagescontent/docs/indice_gazole.xml"
  end

  def time_period
    @parsed_values["titre"]
  end

  def air_multiplier
    format_multiplier @parsed_values["aerien"]
  end

  def road_multiplier
    format_multiplier @parsed_values["routier"]
  end

  private

  def response
    ::HTTP.timeout(10).get(url)
  rescue HTTP::Error
    ""
  end

  def xml_root
    doc = Document.new response.to_s
    doc.root
  end

  def format_multiplier(value)
    return unless value

    value.tr!("%", "")
    value.tr!(",", ".")

    number = (value.to_f / 100 + 1).round(4)
    BigDecimal(number.to_s)
  end
end
